File: 백준/Gold/1987. 알파벳/알파벳.py
from sys import stdin
import sys

r, c = map(int, stdin.readline().split())

matrix = [list(map(str, stdin.readline().rstrip())) for _ in range(r)]
global max_element_count
max_element_count = -1
direct = [[-1, 0], [0, -1], [1, 0], [0, 1]]

def can_go(x, y, matrix):
    return 0 <= x < len(matrix[0]) and 0 <= y < len(matrix)

def dfs(x, y, matrix, element_set):
    global max_element_count
    for i in range(4):
        new_x, new_y = x + direct[i][0], y + direct[i][1]
        if can_go(new_x, new_y, matrix) and matrix[new_y][new_x] not in element_set:
            element_set.add(matrix[new_y][new_x])
            dfs(new_x, new_y, matrix, element_set)
            element_set.remove(matrix[new_y][new_x])

    max_element_count = max(max_element_count, len(element_set))
    if max_element_count == 26:
        print(max_element_count)
        sys.exit(0)

dfs(0, 0, matrix,set(matrix[0][0]))
print(max_element_count)

File: 백준/Gold/1987. 알파벳/알파벳.java
import java.io.*;
import java.util.*;

public class Main {
	static int max;
	static List<List<Integer>> direction = Arrays.asList(Arrays.asList(1,0), Arrays.asList(0,1),
			Arrays.asList(-1,0), Arrays.asList(0,-1)); 
	public static void main(String[] args) throws IOException{
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine()," ");
		
		int height = Integer.parseInt(st.nextToken());
		int width = Integer.parseInt(st.nextToken());
		
		max = 0;
		char[][] matrix = makeMatrix(br, height);
		
		dfs(matrix, 0, 0, new HashSet<>(Arrays.asList(matrix[0][0])));
		System.out.println(max);
	}
	static char[][] makeMatrix(BufferedReader br, int height) throws IOException{
		char[][] matrix = new char[height][];
		for(int i = 0; i < matrix.length; i++) {
			matrix[i] = br.readLine().trim().toCharArray();
		}
		return matrix;
	}
	static void dfs(char[][] matrix, int x, int y, Set<Character> set) {
		for(int i = 0; i < 4; i++) {
			int newX = x + direction.get(i).get(0);
			int newY = y + direction.get(i).get(1);
			if(canGo(newX, newY, matrix) && !set.contains(matrix[newY][newX])) {
				set.add(matrix[newY][newX]);
				dfs(matrix, newX, newY, set);
				set.remove(matrix[newY][newX]);
			}
		}
		
		if(set.size() > max) {
			max = set.size();
		}
	}
	static boolean canGo(int x, int y, char[][] matrix) {
		return (0 <= x && x < matrix[0].length) && (0 <= y && y < matrix.length);
	}
}